### PR TITLE
osutil: allow creating strings out of MountInfoEntry

### PR DESCRIPTION
**Pull request by zyga**
This is mainly done so that testing and debugging is easier as we never
have to write mountinfo records for data storage.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: osutil/mountinfo.go [hunk 1]
```diff
@@ -46,6 +48,49 @@ type MountInfoEntry struct {
...
+		}
+	}
+	return buf.String()
+}
+
+func flattenList(l []string) string {
+	var buf bytes.Buffer
+	for i, el := range l {
+		if i > 0 {
+			fmt.Fprintf(&buf, ",")
```
Comment (bboozzoo) [inline osutil/mountinfo.go]: `buf.WriteString(",")`

File: osutil/mountinfo.go [hunk 2]
```diff
@@ -46,6 +48,49 @@ type MountInfoEntry struct {
...
+		}
+		fmt.Fprintf(&buf, "%s", el)
```
Comment (bboozzoo) [inline osutil/mountinfo.go]: `buf.WriteString(el)`

File: osutil/mountinfo.go [hunk 3]
```diff
@@ -46,6 +48,49 @@ type MountInfoEntry struct {
...
+		if m[key] != "" {
+			fmt.Fprintf(&buf, "%s=%s", key, m[key])
+		} else {
+			fmt.Fprintf(&buf, "%s", key)
+		}
+	}
+	return buf.String()
+}
+
+func flattenList(l []string) string {
```
Comment (bboozzoo) [inline osutil/mountinfo.go]: Why not use `strings.Join(l, ",")`?

File: osutil/mountinfo.go [hunk 4]
```diff
@@ -46,6 +48,49 @@ type MountInfoEntry struct {
...
+func flattenMap(m map[string]string) string {
+	keys := make([]string, 0, len(m))
+	for key := range m {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	var buf bytes.Buffer
+	for i, key := range keys {
+		if i > 0 {
+			fmt.Fprintf(&buf, ",")
```
Comment (bboozzoo) [inline osutil/mountinfo.go]: `buf.WriteString(",")`

File: osutil/mountinfo.go [hunk 5]
```diff
@@ -46,6 +48,49 @@ type MountInfoEntry struct {
...
+		}
+		if m[key] != "" {
+			fmt.Fprintf(&buf, "%s=%s", key, m[key])
+		} else {
+			fmt.Fprintf(&buf, "%s", key)
```
Comment (bboozzoo) [inline osutil/mountinfo.go]: `buf.WriteString(key)`

File: osutil/mountinfo.go (hunk 3 above)
Comment (zyga) [reply osutil/mountinfo.go]: Hmm, because I'm silly :-) EDIT: Will do

File: osutil/mountinfo.go [hunk 6]
```diff
@@ -46,6 +48,42 @@ type MountInfoEntry struct {
...
+func (mi *MountInfoEntry) String() string {
+	maybeSpace := " "
+	if len(mi.OptionalFields) == 0 {
+		maybeSpace = ""
+	}
+	return fmt.Sprintf("%d %d %d:%d %s %s %s %s%s- %s %s %s",
+		mi.MountID, mi.ParentID, mi.DevMajor, mi.DevMinor, escape(mi.Root),
+		escape(mi.MountDir), flattenMap(mi.MountOptions), flattenList(mi.OptionalFields),
+		maybeSpace, mi.FsType, escape(mi.MountSource),
+		flattenMap(mi.SuperOptions))
```
Comment (jdstrand) [inline osutil/mountinfo.go]: You're escaping the strings in this fmt.Sprintf, but aren't in the flattenList and flattenMap functions. Shouldn't you?

File: osutil/mountinfo_test.go [hunk 7]
```diff
@@ -35,9 +35,12 @@ var _ = Suite(&mountinfoSuite{})
 
 // Check that parsing the example from kernel documentation works correctly.
 func (s *mountinfoSuite) TestParseMountInfoEntry1(c *C) {
-	entry, err := osutil.ParseMountInfoEntry(
-		"36 35 98:0 /mnt1 /mnt2 rw,noatime master:1 - ext3 /dev/root rw,errors=continue")
+	real := "36 35 98:0 /mnt1 /mnt2 rw,noatime master:1 - ext3 /dev/root rw,errors=continue"
+	canonical := "36 35 98:0 /mnt1 /mnt2 noatime,rw master:1 - ext3 /dev/root errors=continue,rw"
```
Comment (bboozzoo) [inline osutil/mountinfo_test.go]: Can you add a test case with escaped paths?

Comment (zyga) [reply osutil/mountinfo_test.go]: Done

File: osutil/mountinfo.go (hunk 6 above)
Comment (zyga) [reply osutil/mountinfo.go]: Ah, good catch. Done and tested now.

**Pull request merged**